qiita_db/sql_connection.py

```
@@ -630,3 +632,381 @@ def get_temp_queue(self):
+                    res = None
+                except PostgresError as e:
+                    # Some other error happened during the execution of the
+                    # query, so we need to rollback
+                    self._raise_execution_error(sql, sql_args, e)
+
+                # Store the results of the current query
+                self._results.append(res)
+
+        # wipe out the already executed queries
+        self._queries = []
+
+        return self._results
+
+    @_checker
+    def execute(self):
+        """Executes the transaction
+
+        Returns
+        -------
+        list of DictCursor
+            The results of all the SQL queries in the transaction
+
+        Raises
+        ------
+        RuntimeError
+            If invoked outside a context
+
+        Notes
+        -----
+        If any exception occurs during the execution transaction, a rollback
+        is executed an no changes are reflected in the database.
+        When calling execute, the transaction will never be committed, it will
+        be automatically committed when leaving the context
+
+        See Also
+        --------
+        execute_fetchlast
+        """
+        try:
+            return self._execute()
+        except Exception:
+            self.rollback()
+            raise
+
+    @_checker
+    def execute_fetchlast(self):
+        """Executes the transaction and returns the last result
+
+        This is a convenient function that is equivalent to
+        `self.execute()[-1][0][0]`
+
+        Returns
+        -------
+        object
+            The first value of the last SQL query executed
+
+        See Also
+        --------
+        execute
+        """
+        return self.execute()[-1][0][0]
+
+    @_checker
+    def commit(self):
+        """Commits the transaction and reset the queries
+        Raises
+        ------
+        RuntimeError
+            If invoked outside a context
+        """
+        self._connection.commit()
```

Yup, I think a try/except is need. I'm closing the connection in the the `except` clause and re-raising the error, as I think we don't want to mask these errors...